### PR TITLE
[WPD][ThinLTO]Add cutoff option for WPD

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
This option applies for _import_ WPD (i.e., when `DevirtModule` pass de-virtualizes according to an imported summary, in ThinLTO backend pipeline). It's meant for debugging (e.g., bisection).

### CODE REVIEW

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp [hunk 1]
```diff
@@ -168,6 +168,14 @@ static cl::list<std::string>
                       cl::desc("Prevent function(s) from being devirtualized"),
                       cl::Hidden, cl::CommaSeparated);
 
+/// If this value is other than 0, the devirt module pass will stop
+/// transformation once the total number of devirtualizations reach the cutoff
+/// value.
+static cl::opt<unsigned> WholeProgramDevirtCutoff(
+    "wholeprogramdevirt-cutoff",
+    cl::desc("Max number of devirtualization for devirt module pass"),
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:176]: s/devirtualization/devirtualizations/

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp [hunk 2]
```diff
@@ -1169,6 +1177,10 @@ void DevirtModule::applySingleImplDevirt(VTableSlotInfo &SlotInfo,
       if (!OptimizedCalls.insert(&VCallSite.CB).second)
         continue;
 
+      if (WholeProgramDevirtCutoff.getNumOccurrences() > 0 &&
+          NumSingleImpl >= WholeProgramDevirtCutoff)
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:1181]: NumSingleImpl is a STATISTIC - I'm not sure if it is safe to use this here as I believe the compiler can be built with statistics disabled, in which case this variable would not be correctly tracking the number.

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp [hunk 3]
```diff
@@ -168,6 +168,14 @@ static cl::list<std::string>
                       cl::desc("Prevent function(s) from being devirtualized"),
                       cl::Hidden, cl::CommaSeparated);
 
+/// If this value is other than 0, the devirt module pass will stop
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:171]: I believe the way the code is written, value 0 is not special and will cause no devirts to happen - assuming the option was specified (i.e. getNumOccurrences() > 0). And in fact we probably want a way to say "do 0 devirtualizations".

Comment (minglotus-6) [reply llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp]: I updated the comment, and yes it makes sense to do 0 devirtualization when this option is explicitly set to zero.

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp (hunk 2 above)
Comment (minglotus-6) [reply llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp]: Thanks for the catch! 

I added `static unsigned NumDevirtCalls` to track the number of de-virtualized calls.

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp (hunk 1 above)
Comment (minglotus-6) [reply llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp]: done.

**Review by teresajohnson — APPROVED** [review]

**Pull request merged**